extensions/MongoDbAtlas/MongoDbAtlas/Helpers/MongoDbAtlasSearchHelper.cs

```diff
@@ -0,0 +1,303 @@
+
+    /// <summary>
+    /// Construct helper to interact with atlas and create mappings etc.
+    /// </summary>
+    /// <param name="connection"></param>
+    /// <param name="dbName"></param>
+    public MongoDbAtlasSearchHelper(string connection, string dbName)
+    {
+        var client = new MongoClient(connection);
+        this._db = client.GetDatabase(dbName);
+    }
+
+    /// <summary>
+    /// Get the name of the index to perform a $search aggregation
+    /// </summary>
+    /// <param name="collectionName"></param>
+    /// <returns></returns>
+    public string GetIndexName(string collectionName) => $"searchix_{collectionName}";
+
+    /// <summary>
+    /// Create an ATLAS index and return the id of the index. It also wait for the index to be
+    /// ready, and create the collection if needed.
+    /// </summary>
+    /// <param name="collectionName"></param>
+    /// <param name="embeddingDimension"></param>
+    /// <returns></returns>
+    public async Task<IndexInfo> CreateIndexAsync(string collectionName, int embeddingDimension)
+    {
+        //I need to be able to create index even if collection does not exists
+        //if collection does not exists, create collection and index
+        //if collection does not exists, index does not exists
+        if (!await this.CollectionExistsAsync(collectionName).ConfigureAwait(false))
+        {
+            //index does not exists because collection does not exists
+            await this._db.CreateCollectionAsync(collectionName).ConfigureAwait(false);
+        }
+
+        var status = await this.GetIndexInfoAsync(collectionName).ConfigureAwait(false);
+        if (status.Exists)
+        {
+            //index exists, but we found that status is "does_not_exist" this identify a stale state.
+            if (status.Status == "does_not_exist")
+            {
+                //delete the index and recreate it.
+                await this.DeleteIndicesAsync(collectionName).ConfigureAwait(false);
+            }
+            else
+            {
+                return status;
+            }
+        }
+
+        //now I can create the index.
+        var command = this.CreateCreationCommand(collectionName, embeddingDimension);
+        var result = await this._db.RunCommandAsync<BsonDocument>(command).ConfigureAwait(false);
+        var creationResult = result["indexesCreated"] as BsonArray;
+        if (creationResult == null || creationResult.Count == 0)
+        {
+            return s_falseIndexInfo;
+        }
+
+        return await this.GetIndexInfoAsync(collectionName).ConfigureAwait(false);
+    }
+
+    /// <summary>
+    /// Delete all the indices for a specific collection
+    /// </summary>
+    /// <param name="collectionName"></param>
+    public async Task DeleteIndicesAsync(string collectionName)
+    {
+        var pipeline = new BsonDocument[]
+        {
+            new() {
+                {
+                    "$listSearchIndexes",
```

Comment:
note that this only works starting M10 instances (not on "free forever" M0)

Reply:
I know, do you know a way to get the list of indexes in an M0 instance? The only workaround I know is an option where you avoid the code checking and creating index and you need to create index manually.

Alternatively, I can simply try to create the index and trap exception of index already existing so it can work even in M0.

Reply:
AFAIK that won't work either, since creating SearchIndexes is also only available starting M10...

Reply:
The only solution would be adding an option to the configuration telling the component not to create the indexes, so when the core code of Kernel Memory tells you to create an index, you will simply skip.

Reply:
I personally prefer less hidden "options" - I would rather handle that specific error (both on index check/creation) and log it accordingly. Additionally I would add it as a requirement to the connector Documentation (maybe with details on how to create it manually in case anyone wants to keep using an M0 instance for development).

Reply:
I'll check, I had a situation where the collection was created but the index no, this is the reason why I always check for the index. The UpdateOneAsync can be avoided, even if I thought that CreateIndexAsync() is a method that is rarely called, I'll check as soon as possible, thanks.